sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnosticsLogger.java

```diff
@@ -0,0 +1,125 @@
+        checkNotNull(config, "Argument 'config' must not be null.");
+        this.config = config;
+    }
+
+    /**
+     * Decides whether to log diagnostics for an operation and emits the logs when needed
+     * @param traceContext the Azure trace context
+     * @param diagnosticsContext the Cosmos DB diagnostic context with metadata for the operation
+     */
+    @Override
+    public final void handleDiagnostics(Context traceContext, CosmosDiagnosticsContext diagnosticsContext) {
+        checkNotNull(diagnosticsContext, "Argument 'diagnosticsContext' must not be null.");
+
+        if (shouldLog(diagnosticsContext)) {
+            this.log(diagnosticsContext);
+        }
+    }
+
+    /**
+     * Decides whether to log diagnostics for an operation
+     * @param diagnosticsContext
+     * @return a flag inidcating whether to log the operation or not
+     */
+    protected boolean shouldLog(CosmosDiagnosticsContext diagnosticsContext) {
+
+        if (!diagnosticsContext.hasCompleted()) {
+            return false;
+        }
+
+        if (shouldLogDueToStatusCode(diagnosticsContext.getStatusCode(), diagnosticsContext.getSubStatusCode())) {
+            return true;
+        }
+
+        ResourceType resourceType = ctxAccessor.getResourceType(diagnosticsContext);
+        OperationType operationType = ctxAccessor.getOperationType(diagnosticsContext);
+
+        if (resourceType == ResourceType.Document) {
+            if (pointOperationTypes.contains(operationType)) {
+                if (diagnosticsContext.getDuration().compareTo(this.config.getPointOperationLatencyThreshold()) >= 1) {
+                    return true;
+                }
+            } else {
+                if (diagnosticsContext.getDuration().compareTo(this.config.getFeedOperationLatencyThreshold()) >= 1) {
+                    return true;
+                }
+            }
+        }
+
+        if (diagnosticsContext.getTotalRequestCharge() > this.config.getRequestChargeThreshold()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean shouldLogDueToStatusCode(int statusCode, int subStatusCode) {
+        return statusCode >= 500 || statusCode == 408 || statusCode == 410;
+    }
+
+    /**
+     * Logs the operation. This method can be overridden for example to emit logs to a different target than log4j
+     * @param ctx
+     */
+    protected void log(CosmosDiagnosticsContext ctx) {
+        if (this.shouldLogDueToStatusCode(ctx.getStatusCode(), ctx.getSubStatusCode())) {
+            logger.warn(
+                "Account: {} -> DB: {}, Col:{}, StatusCode: {}:{} Diagnostics: {}",
+                ctx.getAccountName(),
+                ctx.getDatabaseName(),
+                ctx.getCollectionName(),
+                ctx.getStatusCode(),
+                ctx.getSubStatusCode(),
+                ctx.toString());
+        } else {
+            logger.info(
```

Comment:
Should the emitted log level be configurable? I.e. to log it at debug, info or warning level?
Also, since toString() on the context could become CPU intensive (if in future it uses JSON etc) one can check the log level enabled for the logger and if not enabled for info or warning then do not call toString or put the rest of the info on the stack since it is not going to be logged anyway.

Reply:
Made the log level configurable.